src/app/shared/recent-utils.ts

```diff
@@ -0,0 +1,88 @@
+import { ErrorHandler } from '@angular/core';
+import { Context, Space } from 'ngx-fabric8-wit';
+import { Observable, ReplaySubject, Subject } from 'rxjs';
+import { ExtProfile, ProfileService } from '../profile/profile.service';
+
+export const RECENT_LENGTH: number = 8;
+
+export interface RecentData <T extends Context | Space> {
+  data: T[];
+  isSaveRequired: boolean;
+}
+
+export function isContext(context: any): context is Context {
+  return (<Context> context.user !== undefined);
+}
+
+export function isSpace(space: any): space is Space {
+  return (<Space> space.id !== undefined);
+}
+
+export abstract class RecentUtils <T extends Context | Space> {
+
+  protected _recent: Subject<T[]>;
+
+  constructor(
+    protected errorHandler: ErrorHandler,
+    protected profileService: ProfileService
+  ) {
+    this._recent = new ReplaySubject<T[]>(1);
+  }
+
+  get recent(): Observable<T[]> {
+    return this._recent;
+  }
+
+  onBroadcastChanged(changed: T, recent: T[]): RecentData<T> {
+    let index: number = 0;
+    if (isContext(changed)) {
+      index = (recent as Context[]).findIndex((context: Context) => context.name === (changed as Context).name);
+    } else if (isSpace(changed)) {
+      index = (recent as Space[]).findIndex((space: Space): boolean => space.id === (changed as Space).id);
+    }
```

Comment:
Rather than having the abstract class handle this logic by doing a type assertion, could we have the subclasses provide a conversion function to the superclass (via constructor) to map the concrete type to an agnostic identifier/index? ie the subclasses would look like:

```
constructor(
  errorHandler: ErrorHandler,
  profileService: ProfileService
  ... etc
) {
  super(errorHandler, profileService, (ctx: Context): string => ctx.name));
}
```

and


```
constructor(
  errorHandler: ErrorHandler,
  profileService: ProfileService
  ... etc
) {
  super(errorHandler, profileService, (space: Space): string => space.id));
}
```

so that the type signature of this adapter/index function is simply `T => string`.

`onBroadcastChanged` would then start off like:

```
onBroadcastChanged(changed: T, recent: T[]): RecentData<T> {
  let index: number = this.indexFn(changed);
```